roles/confluent.zookeeper/tasks/main.yml

```diff
@@ -14,6 +14,20 @@
     - ansible_os_family
     - ansible_fqdn
 
+- name: Get Package Facts
+  package_facts:
+    manager: auto
+  when: installation_method == "package"
+
+- name: Stop Zookeeper Service if upgrade
+  systemd:
+    name: "{{ zookeeper_service_name }}"
+    state: stopped
+  when:
+    - installation_method == "package"
+    - ansible_facts.packages[zookeeper_main_package] is defined
```

Comment:
This will not stop the service, in case the main package name has changed. It seems like you are trying to avoid fetching too many facts, probably for performance reasons, but another option would be to use the [service_facts module](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/service_facts_module.html) to find out if service exists.

Reply:
Ya its too bad the `package_facts` module doesn't let you filter to specific packages. You are right that it seem unnecessary to gather a bunch of facts when you only want 1. But I think package_facts is fine! I'm not sure we need the `service_facts` module, but maybe. Here are a few scenarios to consider. I think you've got most covered:

First time install:
- ansible_facts.packages[zookeeper_main_package] wont be defined -> this is good, we do not need to stop the nonexistent service

Reconfiguration:
- `ansible_facts.packages[zookeeper_main_package][0].version == confluent_package_version` --> also no need to stop the service

Upgrade (confluent-server):
- `ansible_facts.packages[zookeeper_main_package][0].version != confluent_package_version` --> yes need to stop the service before upgrading packages

Upgrade (confluent-kafka)
- if `ansible_facts.packages['confluent-kafka-2.12] is defined` --> we need to stop the service **AND remove that package** before installing confluent-kafka

Also... another thing i discovered when doing upgrade testing, on debian vs rhel hosts 
`ansible_facts.packages[zookeeper_main_package][0].version` may be either 6.0.1 or 6.0.1-1. A filter could remove that -1 if it exists. When developing the upgrade playbooks i used these conditionals
```
        - kafka_broker_current_version != confluent_full_package_version
        - kafka_broker_current_version != confluent_package_version
```

Why do you think the `service_facts` module is necessary here? That module would be useful here:  https://github.com/confluentinc/cp-ansible/blob/6.1.1-post/kafka_broker.yml#L11

Reply:
Thanks for valuable input. I will "sleep on it", and try to come up with something slightly better by tomorrow.

Reply:
This whole block could be

```
- name: Get Package Facts
  package_facts:
    manager: auto
  when: installation_method == "package"

- name: Stop Zookeeper Service if upgrade
  systemd:
    name: "{{ zookeeper_service_name }}"
    state: stopped
  when: >
    installation_method == "package" and 
    ( ansible_facts.packages[zookeeper_main_package] is defined and ansible_facts.packages[zookeeper_main_package][0].version != confluent_package_version and ansible_facts.packages[zookeeper_main_package][0].version != confluent_full_package_version ) or 
    ansible_facts.packages['confluent-kafka-2.12'] is defined

- name: Remove confluent-kafka-2.12 Package - Red Hat
  yum:
    name: confluent-kafka-2.12
    state: absent
  when:
    - ansible_os_family == "RedHat"
    - ansible_facts.packages['confluent-kafka-2.12'] is defined
    - installation_method == "package"
  # Packages will not be found in check mode because repos not changed
  ignore_errors: "{{ ansible_check_mode }}"

- name: Remove confluent-kafka-2.12 Package - Debian
  apt:
    name: confluent-kafka-2.12
    state: absent
  when:
    - ansible_os_family == "Debian"
    - ansible_facts.packages['confluent-kafka-2.12'] is defined
    - installation_method == "package"
  # Packages will not be found in check mode because repos not changed
  ignore_errors: "{{ ansible_check_mode }}"
```

I think that covers it

Reply:
Yes, that should do the job. A bit afraid of removing the confluent-kafka-2.12 package as part of the desired state. Would be nice to fix https://github.com/confluentinc/cp-ansible/issues/423, but I have to think how it actually can be done - if possible. We are also running into various problems in our DEV environments, where kafka broker and zookeeper are running on the same hosts. We have plans to change that, because of all problems resulting from that configuration.....

Reply:
So the thing about that issue is its the equivalent of upgrading packages on a colocated host. like switching confluent-server-5.5.0 to confluent-server-6.0.0 could have just as bad effects. The solution to that one is to stop both zk and kafka. and then upgrade zk and kafka on that host. then move onto the next.

Or move to the tar install. and then you are safe from this "shared packages problem"

Reply:
Do you think we can ignore that issue here then? I am fine with that.

Reply:
Yes this PR doesn't fully solve that issue, but helps. My solution for that issue is to tell users that colocated hosts must be upgraded at the same time with:
```
ansible-playbook -i hosts.yml all.yml --limit <colocated-host1>
```
The work done in this PR makes all.yml successfully upgrade a component. 

And further when you do all.yml limited to one host it will upgrade zk (leaving kafka in a bad state), and immediately move onto kafka (hopefully fixing that bad state). I think thats the best we can do! Def need to test that use case. 

The only other thing i can think of is inserting additional logic to stop other services running on that host... but I'd rather not. at least for this iteration.

Reply:
@domenicbove PTAL, I think I have covered all the scenarios on your list now. It would be most efficient if we can agree on the pattern for doing this before I add the same logic to the other component roles. And another thing to decide before duplicating this, is to agree on the target branch. Is `6.1.x` ok, or should we target for `6.2.x`? I would prefer `6.1.x`, but I understand if this change is too big to be included in a patch release....

We are still finalizing the upgrade from CP 5.4 to 5.5, as we decided to do this first - before heading for CP 6.x in our clusters still running CP 5.4. One relevant thing I noticed in the process: When adding logic on top of the `confluent_package_version`, as in this PR, we have to make sure it is kept up-to-date. Currently there seems to be a "bug" on the 5.4.x branch: https://github.com/confluentinc/cp-ansible/blob/2def9f2f8a43af08bef1f1c1c463fcf221037983/roles/confluent.variables_handlers/defaults/main.yml#L5 We are running version `5.4.3` in our CP 5.4 clusters. If the upgrade logic in this PR were present on this branch, the services would be stopped on each ansible run.

Another thing to consider, is to support _downgrade_ in the desired state playbooks, but that will be proposed in another round. 😅 We are currently implementing downgrade from CP 5.5 to CP 5.4, and we see at least two relevant use cases:

- Testing upgrade several times, in the same environment. Typically DEV or early TEST environments.
- In case of an emergency rollback after upgrade, either because of hitting a bug in Kafka itself or because clients struggeling for "some reason" (and we need more time to look into it).

Reply:
@erikgb Great work with those changes! I am good with the approach. I also am fine with 6.1.x, but thanks for asking. I agree this is pushing closely toward new feature territory, but since you are doing the work then let's do 6.1.x!

I just updated the 5.4.3-post and 5.4.x branches. Nice catch!

Ya downgrade is interesting. I wonder if you need to yum remove the packages or if things just work. Let's track that elsewhere.

The interesting thing about downgrades is if you go with the "archive" installation method. downgrades are a breeze. just update one line in your override.conf and restart and its downgraded. I also like the archive install because there's no problem with colocating services. You could have zk running 6.1.1 and kafka running 6.0.1 just fine bc the jars/scripts are in separate directories.

Reply:
@domenicbove Thanks! I will update the PR using the same pattern for the other component roles. Please QA carefully; a fairy dies whenever someone copy-modifies code. 😉.

We will discuss downgrade elsewhere, but for the reference, this is how we fixed the downgrade from CP 5.5 to 5.4 (after testing our upgrade):
```yaml
- name: Stop Service
  systemd:
    name: "{{ service_name }}"
    state: stopped

- name: List installed packages
  yum:
    list: installed
  register: yum_installed

- name: Remove all packages starting with confluent-
  yum:
    name: "{{ yum_installed.results | map(attribute='name') | select('search', '^confluent-*') | list }}"
    state: absent
```

Reply:
While downgrade was agreed postponed; looking at the code above (^^), I realize that we could just uninstall all `confluent-` packages if version changed - instead of just uninstalling the `confluent-kafka-2.12` package. Then I think downgrade is supported by this PR as well. WDYT?

Reply:
Ya removing packages also has the benefit of helping you switch from `confluent-kafka` to `confluent-server`. In fact our upgrade docs say to remove all packages. The only reason to go with yum upgrade is potential time savings. I'll do some tests Monday, but I like the idea. 

Also now that i've introduced the `package` tag, skipping these tasks is easy. The package installation tasks hand down take the longest of everything so let's make sure that tag is still functional.

Reply:
Nice! Both `yum` and `apt` ansible modules now support a list supplied to the `name` attribute. That could speed up the package stuff, both install and remove! The loop-variant seems to be considerably slower - I realized this when working on our downgrade tasks. And for removing packages, we need to use a list anyway. If not, the packages list supplied to the loop have to be ordered by dependency. And that I do not want to do....

Reply:
Im totally on board with using a list within the yum module! The only thing is we need to make sure that the package versions work out. I think if the package list variables get updated like this:

```
zookeeper_packages:
  - "confluent-common{{confluent_package_suffix}}"
  - "confluent-rest-utils{{confluent_package_suffix}}"
  - "confluent-metadata-service{{confluent_package_suffix}}"
  - "confluent-ce-kafka-http-server{{confluent_package_suffix}}"
  - "confluent-kafka-rest{{confluent_package_suffix}}"
  - "confluent-server-rest{{confluent_package_suffix}}"
  - "confluent-telemetry{{confluent_package_suffix}}"
  - "{{ zookeeper_main_package }}{{confluent_package_suffix}}"
```

And the package suffix has the proper logic built in
```
confluent_package_suffix: "{{confluent_package_redhat_suffix if ansible_os_family == "RedHat" else confluent_package_debian_suffix}}"
```

Reply:
I think it can be done simpler, and prepared a PR for you: https://github.com/confluentinc/cp-ansible/pull/615. 😉 If the CI-build fails on this PR, I might need your help in tracking down the typo....

Reply:
Hey! So I think you are on to something. This PR could be morphed into containing upgrade **and downgrade** functionality if you just remove all packages when a version changes. I think that will be a huge win. 

Also we could squeeze in the capability to switch from confluent-kafka to confluent-server (and vise versa) here! Another big win for our users